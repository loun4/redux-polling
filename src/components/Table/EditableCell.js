import React from 'react';
import propTypes from 'prop-types';

const EditableCell = ({ value, onChange, onFocus }) => {
  const handleChange = e => onChange(Number(e.target.value));

  return (
    <td>
      <input
        type="number"
        step=".01"
        value={value}
        onChange={handleChange}
        onFocus={onFocus}
      />
    </td>
  );
};

EditableCell.propTypes = {
  value: propTypes.number.isRequired,
  onChange: propTypes.func.isRequired,
  onFocus: propTypes.func,
};

EditableCell.defaultProps = {
  onFocus: null,
};

export default EditableCell;
