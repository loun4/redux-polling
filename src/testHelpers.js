export const wait = async delay =>
  new Promise(resolve => setTimeout(resolve, delay));

export const mockResponse = data => fetch.mockResponse(JSON.stringify(data));

export const data = {
  items: [
    {
      timestamp: 1577205012991,
      index: 1,
      stocks: { NASDAQ: 15.000035664937867, CAC40: 144.95085585450022 },
    },
    {
      timestamp: 1577205013993,
      index: 2,
      stocks: { NASDAQ: 14.999807535966454, CAC40: 144.94473794314257 },
    },
    {
      timestamp: 1577205014994,
      index: 3,
      stocks: { NASDAQ: 14.923148783552527, CAC40: 54.832722903277684 },
    },
    {
      timestamp: 1577205015995,
      index: 4,
      stocks: { NASDAQ: 14.910386463812493, CAC40: 54.83272271392798 },
    },
    {
      timestamp: 1577205016996,
      index: 5,
      stocks: { NASDAQ: 15.784500875221571, CAC40: 53.0103932525638 },
    },
    {
      timestamp: 1577205017997,
      index: 6,
      stocks: { NASDAQ: 15.800954512976684, CAC40: 51.964596975905756 },
    },
    {
      timestamp: 1577205018998,
      index: 7,
      stocks: { NASDAQ: 15.75607527534057, CAC40: 82.92863498029928 },
    },
    {
      timestamp: 1577205019999,
      index: 8,
      stocks: { NASDAQ: 15.755362566834256, CAC40: 79.77210957972518 },
    },
    {
      timestamp: 1577205020999,
      index: 9,
      stocks: { NASDAQ: 17.959520734996243, CAC40: 81.31317656817889 },
    },
    {
      timestamp: 1577205022001,
      index: 10,
      stocks: { NASDAQ: 14.79792027248049, CAC40: 109.72239916988094 },
    },
    {
      timestamp: 1577205023002,
      index: 11,
      stocks: { NASDAQ: 20.6455208726967, CAC40: 110.43443262937416 },
    },
    {
      timestamp: 1577205024003,
      index: 12,
      stocks: { NASDAQ: 32.186997789424034, CAC40: 56.50619716004383 },
    },
    {
      timestamp: 1577205025004,
      index: 13,
      stocks: { NASDAQ: 32.1869992400095, CAC40: 56.49264660421396 },
    },
    {
      timestamp: 1577205026005,
      index: 14,
      stocks: { NASDAQ: 33.073425565636114, CAC40: 63.68522669833696 },
    },
    {
      timestamp: 1577205027006,
      index: 15,
      stocks: { NASDAQ: 33.06895342921482, CAC40: 46.029524704391044 },
    },
    {
      timestamp: 1577205028007,
      index: 16,
      stocks: { NASDAQ: 51.96495250880295, CAC40: 46.029524704084196 },
    },
    {
      timestamp: 1577205029008,
      index: 17,
      stocks: { NASDAQ: 60.48395695643411, CAC40: 19.981267480499998 },
    },
  ],
  last: {
    timestamp: 1577205029008,
    index: 19,
    stocks: { NASDAQ: 60.48395695643412, CAC40: 19.981267480491 },
  },
};

// dirty hack to handle unhandled values returned by the API
// Like 4.0330219586014976e-84 ...
const formatNumber = nbr => {
  const magicSlice = 5;
  const [n1, n2] = String(nbr).split('.');
  return Number(Number(`${n1}.${n2.slice(0, magicSlice)}`).toFixed(2));
};

export const records = data.items.map(({ stocks, index }) => ({
  id: index,
  NASDAQ: formatNumber(stocks.NASDAQ),
  CAC40: formatNumber(stocks.CAC40),
}));
