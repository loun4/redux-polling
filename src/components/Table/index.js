import React, { useRef, useEffect, useCallback } from 'react';
import propTypes from 'prop-types';
import EditableCell from './EditableCell';

const Table = ({ data, editMode, onChange, onBlur, onFocus }) => {
  const tableRef = useRef(null);

  const handleChange = useCallback(
    (field, position) => editValue =>
      onChange({
        field,
        position,
        editValue,
      }),
    [onChange]
  );

  const handleBlur = useCallback(
    e => {
      if (
        editMode &&
        tableRef.current &&
        !tableRef.current.contains(e.target)
      ) {
        onBlur();
      }
    },
    [editMode, onBlur]
  );

  useEffect(() => {
    document.addEventListener('click', handleBlur, false);
    return () => {
      document.removeEventListener('click', handleBlur, false);
    };
  }, [handleBlur]);

  return (
    <table ref={tableRef}>
      <tbody>
        {['NASDAQ', 'CAC40'].map(field => (
          <tr key={field}>
            <th>{field}</th>
            {data.map((item, index) => (
              <EditableCell
                key={item.id}
                value={item[field]}
                onChange={handleChange(field, index)}
                onFocus={editMode ? null : onFocus}
              />
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
};

Table.propTypes = {
  data: propTypes.arrayOf(propTypes.shape({})).isRequired,
  onChange: propTypes.func.isRequired,
  onBlur: propTypes.func.isRequired,
  onFocus: propTypes.func.isRequired,
  editMode: propTypes.bool,
};

Table.defaultProps = {
  editMode: false,
};

export default Table;
