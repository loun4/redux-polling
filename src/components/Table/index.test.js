import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import Table from './index';
import { records } from '../../testHelpers';

const noop = () => {};

const buildSubject = (onBlur = noop, onChange = noop, editMode = false) => {
  return (
    <Table
      data={records}
      editMode={editMode}
      onFocus={noop}
      onChange={onChange}
      onBlur={onBlur}
    />
  );
};

test('Render Table', () => {
  const tree = renderer.create(buildSubject()).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render with NASDAQ & CAC40 rows', () => {
  const wrapper = shallow(buildSubject());
  expect(wrapper.find('tr')).toHaveLength(2);
  expect(
    wrapper
      .find('th')
      .first()
      .text()
  ).toEqual('NASDAQ');
  expect(
    wrapper
      .find('th')
      .last()
      .text()
  ).toEqual('CAC40');
});

test('Render as many EditableCell as data items', () => {
  const wrapper = shallow(buildSubject());
  expect(wrapper.find('EditableCell')).toHaveLength(records.length * 2);
});

test('Call onBlur on editMode', () => {
  const map = {};
  document.addEventListener = jest.fn((event, cb) => {
    map[event] = cb;
  });

  // editMode true
  let mockHandler = jest.fn();
  mount(buildSubject(mockHandler, undefined, true));
  map.click({ target: null });
  expect(mockHandler).toHaveBeenCalledTimes(1);

  // editMode false
  mockHandler = jest.fn();
  mount(buildSubject(mockHandler, undefined, false));
  map.click({ target: null });
  expect(mockHandler).toHaveBeenCalledTimes(0);
});

test('Call onChange with {field, position, editValue}', () => {
  const mockHandler = jest.fn();
  const wrapper = shallow(buildSubject(undefined, mockHandler));

  wrapper
    .find('EditableCell')
    .first()
    .props()
    .onChange(50);

  expect(mockHandler).toHaveBeenCalledWith({
    field: 'NASDAQ',
    position: 0,
    editValue: 50,
  });
});
