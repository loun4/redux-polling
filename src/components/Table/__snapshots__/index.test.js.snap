// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Table 1`] = `
<table>
  <tbody>
    <tr>
      <th>
        NASDAQ
      </th>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={15}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={15}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={14.92}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={14.91}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={15.78}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={15.8}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={15.76}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={15.76}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={17.96}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={14.8}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={20.65}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={32.19}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={32.19}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={33.07}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={33.07}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={51.96}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={60.48}
        />
      </td>
    </tr>
    <tr>
      <th>
        CAC40
      </th>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={144.95}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={144.94}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={54.83}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={54.83}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={53.01}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={51.96}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={82.93}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={79.77}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={81.31}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={109.72}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={110.43}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={56.51}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={56.49}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={63.69}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={46.03}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={46.03}
        />
      </td>
      <td>
        <input
          onChange={[Function]}
          onFocus={[Function]}
          step=".01"
          type="number"
          value={19.98}
        />
      </td>
    </tr>
  </tbody>
</table>
`;
